pkg/elle/txn/cycle.go

```diff
@@ -0,0 +1,214 @@
+}
+
+func fromRelsWithFilter(filter FilterExType, rels ...core.Rel) CycleAnomalySpecType {
+	relsSet := map[core.Rel]struct{}{}
+	for _, v := range rels {
+		relsSet[v] = struct{}{}
+	}
+	return CycleAnomalySpecType{
+		Rels:     relsSet,
+		FilterEx: filter,
+	}
+}
+
+func fromFirstRelAndRest(first core.Rel, rests ...core.Rel) CycleAnomalySpecType {
+	return fromFirstRelAndRestWithFilter(nil, first, rests...)
+}
+
+func fromFirstRelAndRestWithFilter(filter FilterExType, first core.Rel, rests ...core.Rel) CycleAnomalySpecType {
+	restSet := map[core.Rel]struct{}{}
+	for _, v := range rests {
+		restSet[v] = struct{}{}
+	}
+	return CycleAnomalySpecType{
+		FirstRel: first,
+		RestRels: restSet,
+		FilterEx: filter,
+	}
+}
+
+func filterClosure(required string) FilterExType {
+	return func(cycleCase *CycleCase) bool {
+		return cycleCase.Type == required
+	}
+}
+
+func init() {
+	CycleAnomalySpecs = map[string]CycleAnomalySpecType{
+		"G0":            fromRels(core.WW),
+		"G1c":           fromFirstRelAndRest(core.WR, core.WW, core.WR),
+		"G-single":      fromFirstRelAndRest(core.RW, core.WW, core.WR),
+		"G-nonadjacent": fromRels(core.WW, core.WR, core.RW),
+
+		"G2-item":          fromFirstRelAndRestWithFilter(filterClosure("G2-item"), core.RW, core.WR, core.RW, core.WW),
+		"G0-process":       fromRelsWithFilter(filterClosure("G0-process"), core.WW, core.Process),
+		"G1c-process":      fromFirstRelAndRestWithFilter(filterClosure("G1c-process"), core.RW, core.WW, core.WR, core.Process),
+		"G-single-process": fromFirstRelAndRestWithFilter(filterClosure("G-single-process"), core.RW, core.WW, core.WR, core.Process),
+		"G2-item-process":  fromFirstRelAndRestWithFilter(filterClosure("G2-item-process"), core.RW, core.WW, core.WR, core.RW, core.Process),
+
+		// realtime
+
+		"G0-realtime":       fromRelsWithFilter(filterClosure("G0-realtime"), core.WW, core.Realtime),
+		"G1c-realtime":      fromFirstRelAndRestWithFilter(filterClosure("G1c-realtime"), core.WR, core.WW, core.WR, core.Realtime),
+		"G-single-realtime": fromFirstRelAndRestWithFilter(filterClosure("G-single-realtime"), core.RW, core.WW, core.WR, core.Realtime),
+		"G2-item-realtime":  fromFirstRelAndRestWithFilter(filterClosure("G2-item-realtime"), core.RW, core.WW, core.WR, core.Realtime, core.RW),
+	}
+
+	v := CycleAnomalySpecs["G-nonadjacent"]
+	v.With = NonadjacentRW
+	v.FilterPathState = func(p interface{}) bool {
+		// TODO: we need more than one rw edge for this to count; otherwise it's G-single.
+		return true
+	}
+
+	CycleTypeNames = map[string]struct{}{
+		"G-nonadjacent-process":  {},
+		"G-nonadjacent-realtime": {},
+	}
+
+	for k, _ := range CycleAnomalySpecs {
+		CycleTypeNames[k] = struct{}{}
+	}
+
+	UnknownAnomalyTypes = map[string]struct{}{
+		"empty-transaction-graph": {},
+		"cycle-search-timeout":    {},
+	}
+
+	ProcessAnalysisTypes = map[string]struct{}{}
+	RealtimeAnalysisTypes = map[string]struct{}{}
+	for k, _ := range CycleAnomalySpecs {
```

Comment:
use CycleTypeNames not CycleAnomalySpecs.

Reply:
OK